### ISSUE TITLE
Returning correct http header and meaningful error context

### ISSUE THREAD

**Issue by cybersiddhu** [dictybase-playground/kubeless-nodefn#10]
In case of non existent gene name, say something like this, 
` https://betafunc.dictybase.org/genes/name/tbc1d5B`, the error response returns,
```json
{
  "meta": {
    "creator": "kubeless function api"
  },
  "source": {
    "pointer": "/genes/name/tbc1d5B"
  },
  "detail": "Cannot read property 'attributes' of undefined",
  "title": "Cannot read property 'attributes' of undefined",
  "status": 500
}
```
The detail and title at least expected to be informative, for example `The gene name blah blah does not exist in the database`. In addition, the following issues should be addressed,
1. This is not a `HTTP 500` error, instead it should be `404 Not Found`. Please have a read [here](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html). It's preferable to understand and return correct HTTP code for error situations. 
2. The HTTP status code in the header and in the json output should match. In this current buggy implementation, check the following output.
```
HTTP/1.1 200 OK  ==========> HERE
Server: nginx/1.13.12
Date: Thu, 25 Oct 2018 13:00:52 GMT
Content-Type: application/vnd.api+json
Content-Length: 215
Connection: keep-alive
X-Powered-By: kubeless
Access-Control-Allow-Origin: *
Strict-Transport-Security: max-age=15724800; includeSubDomains

{"status":500,"title":"Cannot read property 'attributes' of undefined","detail":"Cannot read property 'attributes' of undefined","source":{"pointer":"/genes/name/tbc1d5B"},"meta":{"creator":"kubeless function api"}}
```
Now, retuning success `200` also caching empty response in the redis which is apparently another side effect of this shortcoming.

**Comment by erichartline** [comment]
The issue here was that the `geneHandler` and `geneNameHandler` functions were not checking if their calls to `gene2name` returned an error or not. I updated them to check if the response had an error; if they did, then it would set the status to match that error code and provide the structured JSON error response. If there is no error, then it provides the expected data.

There was also an issue with Redis setting a cache any time the user hit a route for the first time, even if the route was for a nonexistent gene. I added a check to see if the data response was an error; if so, then it doesn't set anything in the cache.

I also added more detailed loggers throughout the different functions to make it easier to debug.
